bpfd-dev/bpfd
# Implement bpffs support in the bpfd CSI plugin

This is the POC skeleton which allows k8s pods to request certain loaded maps from bpfd. For example with it you can run 

```bash
cat <<EOF | kubectl apply -f -
apiVersion: bpfd.dev/v1alpha1
kind: XdpProgram
metadata:
  labels:
    app.kubernetes.io/name: xdpprogram
  name: go-xdp-counter-example
spec:
  sectionname: xdp_stats
  # Select all nodes
  nodeselector: {}
  interfaceselector:
    primarynodeinterface: true
  priority: 55
  bytecode:
    image:
      url: quay.io/bpfd-bytecode/go-xdp-counter:latest
---
apiVersion: v1
kind: Pod
metadata:
  name: test
spec:
  containers:
  - name: main
    image: nginx
    volumeMounts:
    - name: bpf-maps
      mountPath: /run/bpfd/fs/maps
      readOnly: true
   volumes:
   - name: bpf-maps
     csi:
       driver: csi.bpfd.dev
       volumeAttributes:
         csi.bpfd.dev/program: go-xdp-counter-example
         csi.bpfd.dev/maps: xdp_stats_map
EOF
```

Which will load a program (xdp-counter) which creates an `xdp_stats_map` and then startup a pod which requests for that map to be mounted in its /run/bpfd/fs for future use.

There's still some TODOs which need to be completed following this PR:

- Move all aya API calls + bpffs creation into the main bpfManager thread which is where any/all privileged operations should take place
- Make this configuration (csi) the default and update all example binaries + deployments to make use of it
- Ensure K8s deployment bpfd privileges are properly fixed.
